lib/auth/auth.go

```diff
@@ -4323,6 +4352,45 @@ func (a *Server) UpsertWindowsDesktop(ctx context.Context, desktop types.Windows
 	return nil
 }
 
+// desktopsLimitExceeded checks if number of non-AD desktops exceeds limit for OSS distribution. Returns always false for Enterprise.
+// If there are more non-AD desktops than allowed it also adds cluster alert to notify user.
+func (a *Server) desktopsLimitExceeded(ctx context.Context) (bool, error) {
+	if modules.GetModules().BuildType() != modules.BuildOSS {
+		return false, nil
+	}
+	startKey := backend.ExactKey(local.WindowsDesktopsPrefix)
+	stream := backend.StreamRange(ctx, a.bk, startKey, backend.RangeEnd(startKey), apidefaults.DefaultChunkSize)
```

Comment:
This will read from the actual backend even though windows desktops are cached; make yourself a helper that uses `api/internalutils/stream.PageFunc` with `a.ListWindowsDesktops` instead, so you don't even need to deal with the unmarshaling.

Completely untested:
```go
func (a *Server) StreamWindowsDesktops(ctx context.Context, startKey string) stream.Stream[types.WindowsDesktop] {
	var done bool
	return stream.PageFunc(func() ([]types.WindowsDesktop, error) {
		if done {
			return nil, io.EOF
		}
		resp, err := a.ListWindowsDesktops(ctx, types.ListWindowsDesktopsRequest{
			Limit:    apidefaults.DefaultChunkSize,
			StartKey: startKey,
		})
		if err != nil {
			return nil, trace.Wrap(err)
		}
		startKey = resp.NextKey
		done = startKey == ""
		return resp.Desktops, nil
	})
}
```

Reply:
I'd also use stream.FilterMap:
```go
stream := stream.FilterMap(
	a.StreamWindowsDesktops(ctx, ""),
	func(d types.WindowsDesktop) (struct{}, bool) {
		return struct{}, desktop.NonAD()
	},
)
```
then your `for` loop can just `count++` and stop the stream as needed.

Reply:
(`DefaultChunkSize` is 1000, it can probably be made to be 50 or 100 since we're reading from the cache in the happy path, so latency isn't an issue, overreading and "overunmarshaling" might be)

Reply:
Thanks! I've changed it as suggested